libz-rs-sys/src/gz.rs

```diff
@@ -0,0 +1,230 @@
+pub use zlib_rs::c_api::*;
+
+use core::ffi::{c_char, c_int, c_uint, c_void, CStr };
+
+use std::ptr;
+use libc::{O_APPEND, O_CLOEXEC, O_CREAT, O_EXCL, O_LARGEFILE, O_RDONLY, O_TRUNC, O_WRONLY, SEEK_CUR, SEEK_END};
+use zlib_rs::deflate::Strategy;
+
+// For compatibility with the zlib C API, this structure exposes just enough of the
+// internal state of an open gzFile to support the gzgetc() C macro.
+#[repr(C)]
+#[derive(Copy, Clone)]
+pub struct gzFile_s {
+    have: c_uint,        // number of bytes available at next
+    next: *const Bytef,  // next byte of uncompressed data
+    pos: u64,            // current offset in uncompressed data stream
+}
+
+// File handle for an open gzip file.
+pub type gzFile = *mut gzFile_s;
+
+// The internals of a gzip file handle (the thing gzFile actually points to, with the
+// public gzFile_s part at the front for ABI compatibility).
+#[repr(C)]
+struct gz_state {
+    x: gzFile_s,         // "x" for the exposed part of the data structure
+
+    // Fields used for both reading and writing
+    mode: GzMode,
+    fd: c_int,           // file descriptor
+    path: *const c_char,
+    size: usize,         // buffer size; can be 0 if not yet allocated
+    want: usize,         // requested buffer size
+    input: *mut Bytef,   // input buffer
+    output: *mut Bytef,  // output buffer
+    direct: bool,        // true in pass-through mode, false if processing gzip data
+
+    // Fields used just for reading
+    // FIXME: add the 'how' field when read support is implemented
+    start: i64,
+    eof: bool,           // whether we have reached the end of the input file
+    past: bool,          // whether a read past the end has been requested
+
+    // Fields used just for writing
+    level: i8,
+    strategy: Strategy,
+    reset: bool,         // whether a reset is pending after a Z_FINISH
+
+    // Fields used for seek requests
+    skip: i64,           // amount to skip (already rewound if backwards)
+    seek: bool,          // whether a seek request is pending
+
+    // Error information
+    err: c_int,          // last error (0 if no error)
+    msg: *const c_char,  // error message from last error (NULL if none)
+
+    // FIXME: add the zstream field when read/write support is implemented
+}
+
+// Gzip operating modes
+// NOTE: These values match what zlib-ng uses.
+#[derive(PartialEq)]
+enum GzMode {
+    GZ_NONE = 0,
+    GZ_READ = 7247,
+    GZ_WRITE = 31153,
+    GZ_APPEND = 1,
+}
+
+const GZBUFSIZE: usize = 128 * 1024;
+
+const Z_DEFAULT_COMPRESSION: i8 = -1;
+
+/// Open a gzip file for reading or writing.
+///
+/// # Safety
+/// The caller must ensure that path and mode point to valid C strings. If the
+/// return value is non-NULL, caller must delete it using only gzclose.
```

Comment:
```suggestion
/// return value is non-NULL, caller must delete it using only [`gzclose`].
```

this makes the function name clickable in the documentation.